Final/RFC00XX-Custom-Remote-Connections.md

```diff
@@ -0,0 +1,281 @@
+        public override void CreateAsync()
+        {
+            // Create writer for this connection.
+            stdInWriter = new TransportTextWriter(_connectionInfo.messageWriter);
+
+            // Start reader thread for this connection.
+            StartReaderThread(_connectionInfo.messageReader);
+        }
+    }
+}
+```
+
+## Alternate Design Options
+
+### PowerShell Subsystem
+
+Another way to implement a custom connection, other than as a PowerShell module, is as a PowerShell subsystem.
+PowerShell subsystems are an experimental feature where external feature implementations can be registered with PowerShell.
+A registered subsystem can integrate with PowerShell via a public interface, but also define cmdlets that become available in the session.  
+
+This proposal does not use the engine subsystem infrastructure because it was deemed unnecessary.
+Implementing custom connections as a PowerShell module is sufficient, and subsystems provide no additional value.
+
+## Background Information
+
+### PowerShell Remoting
+
+PowerShell remoting consists of a number of internal engine components.
+It provides a way to establish a PowerShell session on a remote target, which can be another physical machine, virtual machine, or local process.
+Commands and scripts can then be run in the remote session, with data objects being returned to the client.
+PowerShell manipulates object instances and so returned data are objects that are serialized when transported over the remoting connection, and then deserialized after arriving at the destination.
+If the object type is known, PowerShell will deserialize it back into its original type.
+Unknown types are deserialized into a general PowerShell `PSCustomObject` type, that is essentially a property bag.  
+
+The core part of PowerShell remoting is the remoting protocol (PSRP) implementation, which handles all messages and data over the connection.
+But there is also the need to create the connections, along with mechanisms to transport protocol messages/data over the connection.
+These functions are considered to be outside PowerShell core remoting, and specific to each connection type.
+PowerShell provides abstractions for these functions, but leaves the implementation to each connection type.  
+
+#### Example - WinRM remoting
+
+This is the first remoting connection supported in PowerShell 2.0 and it is based on Windows WinRM.
+WinRM is Windows secure remote shell solution.
+PowerShell remoting uses WinRM to establish a remote connection, and then create a PowerShell session on a remote target.
+PowerShell uses the WinRM plugin architecture to create PowerShell target endpoint sessions as WinRM remote shell instances.
+The PowerShell engine provides WinRM specific connection objects, `WSManConnectionInfo`, to create a connection through WinRM.
+It also contains a transport implementation for transferring protocol messages and data through the WinRM connection.
+
+#### Example - SSH remoting
```

Comment:
I'd like to add that SSH remoting is by far the most "decoupled" PowerShell remoting transport in the sense that it it does everything with standard input/output streams on both sides - there is really nothing much that is directly tied to OpenSSH in particular beyond the fact that it expects to call ssh.exe with specific command-line arguments. This is the transport type that has a lot of potential for generalization and easy integration with unmanaged languages like C/Rust using an external executable that behaves like OpenSSH.

Reply:
Eh, most of the OutOfProc ones are like this. SSH just uses the stdio ports of the process to communicate but the named pipe stuff is a bidirectional pipe. The proposal here is to allow people to implement their own connections and the mechanism for exchanging data is using 2 .NET streams leading to the same thing really.